File: Salary Prediction_Data Science and Predictive Modeling/scripts/r/01_EDA.R
# Exploratory Data Analysis (EDA) Script - Position Salaries Dataset
# This script performs comprehensive EDA on the Position Salaries dataset.

# Load necessary libraries
library(tidyverse)
library(ggplot2)
library(dplyr)
library(readr)
library(corrplot)

# Set working directory
project_root <- dirname(dirname(dirname(getwd())))
data_path <- file.path(project_root, "data", "raw", "Position_Salaries.csv")

# Load data
df <- read_csv(data_path)

# Create output directory
output_dir <- file.path(project_root, "results", "figures")
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

# Data Overview
cat("=" , rep("=", 60), "\n", sep = "")
cat("DATA OVERVIEW\n")
cat("=" , rep("=", 60), "\n", sep = "")
cat("\nDataset Shape:", dim(df), "\n")
cat("\nFirst few rows:\n")
print(head(df, 10))
cat("\nDataset Summary:\n")
print(summary(df))
cat("\nMissing Values:\n")
print(colSums(is.na(df)))
cat("\nDuplicate rows:", sum(duplicated(df)), "\n")

# Statistical Summary
cat("\nStatistical Summary:\n")
print(summary(df))

# Visualizations
# 1. Salary Distribution
png(file.path(output_dir, "salary_distribution.png"), width = 1200, height = 600)
p1 <- ggplot(df, aes(x = Salary)) +
  geom_histogram(bins = 10, fill = "skyblue", color = "black", alpha = 0.7) +
  labs(title = "Distribution of Salary", x = "Salary", y = "Frequency") +
  theme_minimal() +
  theme(plot.title = element_text(size = 14, face = "bold"))
print(p1)
dev.off()

# 2. Level vs Salary
png(file.path(output_dir, "level_vs_salary.png"), width = 1200, height = 600)
p2 <- ggplot(df, aes(x = Level, y = Salary)) +
  geom_point(size = 3, color = "coral", alpha = 0.7) +
  geom_line(color = "green", linewidth = 1) +
  labs(title = "Level vs Salary", x = "Level", y = "Salary") +
  theme_minimal() +
  theme(plot.title = element_text(size = 14, face = "bold"))
print(p2)
dev.off()

# 3. Salary by Position
png(file.path(output_dir, "salary_by_position.png"), width = 1400, height = 800)
df_sorted <- df %>% arrange(Salary)
p3 <- ggplot(df_sorted, aes(x = reorder(Position, Salary), y = Salary)) +
  geom_bar(stat = "identity", fill = "steelblue", alpha = 0.8) +
  coord_flip() +
  labs(title = "Salary by Position", x = "Position", y = "Salary ($)") +
  theme_minimal() +
  theme(plot.title = element_text(size = 16, face = "bold")) +
  geom_text(aes(label = paste0("$", format(Salary, big.mark = ","))), 
            hjust = -0.1, size = 3)
print(p3)
dev.off()

# 4. Correlation
correlation <- cor(df$Level, df$Salary)
cat("\nCorrelation (Level vs Salary):", correlation, "\n")

# Save processed data
processed_path <- file.path(project_root, "data", "processed", "processed_data.csv")
dir.create(dirname(processed_path), recursive = TRUE, showWarnings = FALSE)
write_csv(df, processed_path)

cat("\nEDA completed successfully!\n")
cat("Visualizations saved to:", output_dir, "\n")
cat("Processed data saved to:", processed_path, "\n")

File: Salary Prediction_Data Science and Predictive Modeling/scripts/python/01_EDA.py
"""
Exploratory Data Analysis (EDA) Script - Position Salaries Dataset
This script performs comprehensive EDA on the Position Salaries dataset.
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from pathlib import Path
import sys

# Add project root to path
project_root = Path(__file__).resolve().parent.parent.parent
sys.path.append(str(project_root))

# Set style and warnings
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (12, 6)
warnings.filterwarnings('ignore')

def load_data():
    """Load the dataset from the raw data directory."""
    data_path = project_root / "data" / "raw" / "Position_Salaries.csv"
    df = pd.read_csv(data_path)
    return df

def data_overview(df):
    """Display basic overview of the dataset."""
    print("=" * 60)
    print("DATA OVERVIEW")
    print("=" * 60)
    print(f"\nDataset Shape: {df.shape}")
    print(f"\nFirst few rows:")
    print(df.head(10))
    print(f"\nDataset Info:")
    print(df.info())
    print(f"\nDataset Description:")
    print(df.describe())
    print(f"\nMissing Values:")
    print(df.isnull().sum())
    print(f"\nDuplicate rows: {df.duplicated().sum()}")

def create_visualizations(df, output_dir):
    """Create and save all visualizations."""
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    
    # 1. Salary Distribution
    plt.figure(figsize=(12, 6))
    plt.subplot(1, 2, 1)
    plt.hist(df['Salary'], bins=10, edgecolor='black', alpha=0.7, color='skyblue')
    plt.title('Distribution of Salary', fontsize=14, fontweight='bold')
    plt.xlabel('Salary', fontsize=12)
    plt.ylabel('Frequency', fontsize=12)
    plt.grid(True, alpha=0.3)
    
    plt.subplot(1, 2, 2)
    plt.boxplot(df['Salary'], vert=True)
    plt.title('Box Plot of Salary', fontsize=14, fontweight='bold')
    plt.ylabel('Salary', fontsize=12)
    plt.grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.savefig(output_dir / 'salary_distribution.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # 2. Level vs Salary
    plt.figure(figsize=(12, 6))
    plt.subplot(1, 2, 1)
    plt.scatter(df['Level'], df['Salary'], s=100, alpha=0.7, color='coral', edgecolors='black')
    plt.title('Level vs Salary', fontsize=14, fontweight='bold')
    plt.xlabel('Level', fontsize=12)
    plt.ylabel('Salary', fontsize=12)
    plt.grid(True, alpha=0.3)
    
    plt.subplot(1, 2, 2)
    plt.plot(df['Level'], df['Salary'], marker='o', linewidth=2, markersize=8, color='green')
    plt.title('Level vs Salary (Line Plot)', fontsize=14, fontweight='bold')
    plt.xlabel('Level', fontsize=12)
    plt.ylabel('Salary', fontsize=12)
    plt.grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.savefig(output_dir / 'level_vs_salary.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # 3. Salary by Position
    plt.figure(figsize=(14, 8))
    df_sorted = df.sort_values('Salary')
    plt.barh(df_sorted['Position'], df_sorted['Salary'], color='steelblue', alpha=0.8)
    plt.title('Salary by Position', fontsize=16, fontweight='bold')
    plt.xlabel('Salary ($)', fontsize=12)
    plt.ylabel('Position', fontsize=12)
    plt.grid(True, alpha=0.3, axis='x')
    
    for i, v in enumerate(df_sorted['Salary']):
        plt.text(v + 10000, i, f'${v:,}', va='center', fontsize=10)
    
    plt.tight_layout()
    plt.savefig(output_dir / 'salary_by_position.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # 4. Correlation Heatmap
    correlation = df[['Level', 'Salary']].corr()
    plt.figure(figsize=(8, 6))
    sns.heatmap(correlation, annot=True, cmap='coolwarm', center=0, 
                square=True, linewidths=2, cbar_kws={"shrink": 0.8})
    plt.title('Correlation Heatmap: Level vs Salary', fontsize=14, fontweight='bold')
    plt.tight_layout()
    plt.savefig(output_dir / 'correlation_heatmap.png', dpi=300, bbox_inches='tight')
    plt.close()

def main():
    """Main function to run EDA."""
    # Load data
    df = load_data()
    
    # Data overview
    data_overview(df)
    
    # Create visualizations
    output_dir = project_root / 'results' / 'figures'
    create_visualizations(df, output_dir)
    
    # Save processed data
    processed_path = project_root / 'data' / 'processed' / 'processed_data.csv'
    processed_path.parent.mkdir(parents=True, exist_ok=True)
    df.to_csv(processed_path, index=False)
    
    print(f"\nEDA completed successfully!")
    print(f"Visualizations saved to: {output_dir}")
    print(f"Processed data saved to: {processed_path}")

if __name__ == "__main__":
    main()


